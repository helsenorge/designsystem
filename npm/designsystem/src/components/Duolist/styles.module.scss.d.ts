export type Styles = {
  duolist: string;
  duolist__dd: string;
  'duolist__dd--bold': string;
  'duolist__dd--non-formatted': string;
  duolist__dt: string;
  'duolist__dt--bold': string;
  'duolist__dt--non-formatted': string;
  'duolist--collapsed': string;
  'duolist--line': string;
  'duolist--non-formatted': string;
  'duolist--not-collapsed': string;
  'duolist-wrapper--border': string;
  'duolist-wrapper--extra-padding-top': string;
};

export type ClassNames = keyof Styles;

declare const styles: Styles;

export default styles;
