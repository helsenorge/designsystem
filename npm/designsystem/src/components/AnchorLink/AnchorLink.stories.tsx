import React from 'react';

import { action } from '@storybook/addon-actions';
import { Meta, StoryObj } from '@storybook/react';

import AnchorLink from './AnchorLink';

const meta = {
  title: '@helsenorge∕designsystem-react/Components/AnchorLink',
  component: AnchorLink,
  parameters: {
    docs: {
      description: {
        component:
          'AnchorLink kan bli brukt som et komponent for \\<a\\> tag eller \\<button\\> tag som skal ha en anchorlink styling.<br>.anchorlink-wrapper kan også bli brukt når det ikke er kontroll over markup som kommer inn.',
      },
    },
  },
  args: {
    children: 'Anchorlink tekst',
    href: '/test',
  },
  argTypes: {
    children: {
      control: 'text',
    },
    href: {
      control: 'text',
    },
  },
} satisfies Meta<typeof AnchorLink>;

export default meta;

type Story = StoryObj<typeof meta>;

export const Default: Story = {
  render: args => (
    <>
      <AnchorLink {...args} target={'_self'}>
        {args.children}
      </AnchorLink>
      <br />
      <br />
      <AnchorLink {...args} target={'_self'}>
        {
          'Eiusmod veniam reprehenderit dolore magna tempor dolor reprehenderit reprehenderit ullamco sit in nulla qui. (Lang tekst - Skal wrappe).'
        }
      </AnchorLink>
    </>
  ),
};

export const External: Story = {
  render: args => (
    <>
      <AnchorLink {...args} target={'_blank'}>
        {args.children}
      </AnchorLink>
      <br />
      <br />
      <AnchorLink {...args} target={'_blank'}>
        {
          'Eiusmod veniam reprehenderit dolore magna tempor dolor reprehenderit reprehenderit ullamco sit in nulla qui. (Lang tekst - Skal wrappe).'
        }
      </AnchorLink>
    </>
  ),
};

export const AsButton: Story = {
  render: args => (
    <p style={{ fontSize: '1.25rem' }}>
      Dette er først en{' '}
      <AnchorLink htmlMarkup={'a'} onClick={action('AnchorLink clicked!')} {...args}>
        vanlig lenke i løpende tekst
      </AnchorLink>{' '}
      og nå kommer en
      <AnchorLink htmlMarkup={'button'} onClick={action('AnchorLink clicked!')} {...args}>
        button-lenke i løpende tekst som går over flere linjer
      </AnchorLink>{' '}
      og til slutt en{' '}
      <AnchorLink htmlMarkup={'button'} onClick={action('AnchorLink clicked!')} {...args}>
        kort
      </AnchorLink>{' '}
      button-lenke
    </p>
  ),
};
