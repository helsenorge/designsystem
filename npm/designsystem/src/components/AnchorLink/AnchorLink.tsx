import React from 'react';
import Icon, { IconSize } from '../Icons';
import ArrowUpRight from '../Icons/ArrowUpRight';
import { getColor } from '../../theme/currys';
import { useHover } from '../../hooks/useHover';

import AnchorLinkStyles from './styles.module.scss';
import classNames from 'classnames';

export type AnchorLinkTargets = '_self' | '_blank' | '_parent';

interface AnchorLinkProps {
  /** Sets the content of the <a> tag */
  children: React.ReactNode;
  /** URL to link to */
  href: string;
  /** Gives a unique id to the anchor-link :) */
  id?: string;
  /** Adds custom classes to the element. */
  className?: string;
  /** Sets the target type of the <a> tag. _blank adds an arrow icon at the end of the link */
  target?: AnchorLinkTargets;
}

const AnchorLink = React.forwardRef(function AnchorLinkForwardedRef(props: AnchorLinkProps, ref: React.ForwardedRef<HTMLAnchorElement>) {
  const { id, href, children, className, target = '_self' } = props;
  const external = target === '_blank';
  const { hoverRef, isHovered } = useHover<HTMLAnchorElement>(ref as React.RefObject<HTMLAnchorElement>, undefined, false);
  const anchorClasses = classNames(AnchorLinkStyles.anchorlink, className ?? '');

  return (
    <a id={id} href={href} className={anchorClasses} target={target} ref={hoverRef}>
      {children}
      {external && (
        <Icon
          className={AnchorLinkStyles.anchorlink__icon}
          svgIcon={ArrowUpRight}
          color={getColor('blueberry', 600)}
          hoverColor={getColor('blueberry', 700)}
          size={IconSize.XSmall}
          isHovered={isHovered}
        />
      )}
    </a>
  );
});

export default AnchorLink;
