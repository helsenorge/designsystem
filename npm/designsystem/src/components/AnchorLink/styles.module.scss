@use 'sass:map';
@import '../../scss/spacers';
@import '../../scss/breakpoints';
@import '../../scss/palette';

$underline-width: 0.0625rem;
$focus-width: 0.15rem;

@mixin anchorlink-hover {
  cursor: pointer;
  color: $blueberry700;
  background-color: $blueberry60010;
  text-decoration-color: $blueberry600;
}

@mixin anchorlink-focus {
  outline: none;
  border: $focus-width solid $black;
}

@mixin anchorlink {
  display: inline;
  box-decoration-break: clone;
  overflow-wrap: break-word;
  word-break: break-word;
  text-align: left;
  color: $blueberry600;
  border: $focus-width solid transparent;
  text-decoration: underline;
  text-underline-offset: 0.16rem;
  text-decoration-color: $blueberry400;
  text-decoration-thickness: $underline-width;
  padding: 0.1rem;

  &:hover {
    @include anchorlink-hover;
  }

  &:focus {
    @include anchorlink-focus;
  }

  &__icon {
    display: inline-block;
    vertical-align: bottom;
    margin-right: -0.4rem;
    margin-top: -2rem;
    margin-bottom: -0.3rem;
  }
}

@mixin anchorlink-icon {
  &[target='_blank'] {
    &::after {
      content: url("data:image/svg+xml;utf8,<svg xmlns='http://www.w3.org/2000/svg'  width='38px' height='38px' viewBox='0 0 48 48' fill='%2308667c'><polygon points='33.305,14.659,33.305,29.205,32.005,29.205,32.005,16.948,15.804,33.149,14.885,32.229,31.155,15.959,18.761,15.96,18.761,14.659' /></svg>");
      display: inline-block;
      vertical-align: bottom;
      margin-right: -0.4rem;
      margin-top: -2rem;
      margin-bottom: -0.8rem;
    }
  }
}

.anchorlink {
  @include anchorlink;
}

.anchorlink-wrapper {
  all: unset;

  &:hover > span {
    @include anchorlink-hover;
  }

  &:focus > span {
    @include anchorlink-focus;
  }
}
