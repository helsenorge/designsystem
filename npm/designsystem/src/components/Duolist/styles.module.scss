@use 'sass:map';
@import '../../scss/breakpoints';
@import '../../scss/font-settings';
@import '../../scss/palette';
@import '../../scss/spacers';

.duolist-wrapper {
  &--border {
    padding: getSpacer(s) getSpacer(s) getSpacer(l);
    border: 0.062rem solid $neutral500;
  }

  &--extra-padding-top {
    padding-top: getSpacer(xl);
    padding-bottom: getSpacer(xl);
  }
}

.duolist {
  font-size: $font-size-sm;
  margin: 0;
  display: grid;

  @media (min-width: map.get($grid-breakpoints, sm)) {
    font-size: $font-size-md;
  }

  &__dt {
    padding-right: getSpacer(l);

    @media (min-width: map.get($grid-breakpoints, sm)) {
      padding-right: 2.6669rem;
    }

    &--non-formatted {
      padding-right: 0;
    }
    &--non-formatted#{&}::after {
      content: attr(data-separator);
      font-weight: 400;
    }
  }

  &__dd {
    margin-inline-start: 0;
  }

  &__dt,
  &__dd {
    line-height: $lineheight-size-md;
    padding-bottom: getSpacer(2xs);

    &:last-of-type {
      padding-bottom: 0;
    }

    &--bold {
      font-weight: 600;
      line-height: $lineheight-size-sm;
    }

    &--non-formatted {
      display: inline;
    }
  }

  &--line &__dt,
  &--line &__dd {
    padding-top: getSpacer(2xs);
    padding-bottom: getSpacer(2xs);
    border-top: 0.062rem solid $neutral200;

    &:last-of-type {
      padding-bottom: getSpacer(2xs);
      border-bottom: 0.062rem solid $neutral200;
    }
  }

  &--non-formatted {
    display: unset;
  }
}
