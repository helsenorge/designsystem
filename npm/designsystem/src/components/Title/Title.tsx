import React from 'react';

import classNames from 'classnames';

import { AnalyticsId } from '../../constants';

import titleStyles from './styles.module.scss';

export type TitleTags = 'h1' | 'h2' | 'h3' | 'h4' | 'h5' | 'h6' | 'span';
export type TitleAppearances = 'titleFeature' | 'title1' | 'title2' | 'title3' | 'title4' | 'title5' | 'title6';

export interface TitleProps {
  children: React.ReactNode;
  /** Gives a unique id to the title */
  id?: string;
  /** Adds custom classes to the element. */
  className?: string;
  /** Adds top and bottom margin in rem. */
  margin?: number | TitleMargin;
  /** Changes the underlying element of the title. */
  htmlMarkup?: TitleTags;
  /** Changes the appearance of the title. */
  appearance?: TitleAppearances;
  /** Sets the data-testid attribute. */
  testId?: string;
}

export const Title = React.forwardRef(function TitleForwardedRef(props: TitleProps, ref: React.ForwardedRef<HTMLHeadingElement>) {
  const { id, children, className, htmlMarkup = 'h1', appearance = 'title1', margin = 0, testId } = props;
  const titleClasses = classNames(
    titleStyles.title,
    {
      [titleStyles['title--feature']]: appearance === 'titleFeature',
      [titleStyles['title--title1']]: appearance === 'title1',
      [titleStyles['title--title2']]: appearance === 'title2',
      [titleStyles['title--title3']]: appearance === 'title3',
      [titleStyles['title--title4']]: appearance === 'title4',
      [titleStyles['title--title5']]: appearance === 'title5',
      [titleStyles['title--title6']]: appearance === 'title6',
    },
    className
  );
  const CustomTag = htmlMarkup;

  const inlineStyle = instanceOfTitleMargin(margin)
    ? { marginTop: `${margin.marginTop}rem`, marginBottom: `${margin.marginBottom}rem` }
    : { marginTop: `${margin}rem`, marginBottom: `${margin}rem` };

  return (
    <CustomTag id={id} className={titleClasses} style={inlineStyle} ref={ref} data-testid={testId} data-analyticsid={AnalyticsId.Title}>
      {children}
    </CustomTag>
  );
});

export interface TitleMargin {
  marginTop: number;
  marginBottom: number;
}

export const instanceOfTitleMargin = (margin: unknown): margin is TitleMargin => {
  return Object.prototype.hasOwnProperty.call(margin, 'marginTop') && Object.prototype.hasOwnProperty.call(margin, 'marginBottom');
};

export default Title;
