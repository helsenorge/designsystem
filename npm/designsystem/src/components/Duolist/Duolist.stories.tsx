import React from 'react';

import { StoryObj, Meta } from '@storybook/react';

import Duolist, { DuolistGroup } from './Duolist';
import { shortLoremText, mediumLoremText } from '../../utils/loremtext';
import AnchorLink from '../AnchorLink';
import Title from '../Title';

const meta = {
  title: '@helsenorge∕designsystem-react/Components/Duolist',
  component: Duolist,
  parameters: {
    docs: {
      description: {
        component:
          'Som innbygger vil jeg kunne se utlisting av parvise dataelementer bestående av parameternavn og parameterverdi ("name/value pairs") slik at jeg kan scanne listen raskt for å se hvilke verdier som gjelder for hvlke elementer.',
      },
    },
  },
  args: {
    children: '',
  },
  argTypes: {
    descriptionWidth: {
      control: 'number',
    },
  },
} satisfies Meta<typeof Duolist>;

export default meta;

type Story = StoryObj<typeof meta>;

export const Default: Story = {
  render: args => (
    <Duolist {...args}>
      <DuolistGroup term={'test term'} description={'kort tekst'} />
      <DuolistGroup term={'test term lang'} description={shortLoremText} />
      <DuolistGroup term={'test term enda lenger'} description={mediumLoremText} />
      <DuolistGroup term={'test term'} description={'test description'} />
    </Duolist>
  ),
};

const descriptionWithLink = (
  <>
    {'Her er en link: '} <AnchorLink href={'https://www.helsenorge.no'}>{'Anchorlink test'}</AnchorLink>
  </>
);

export const DescriptionWithLink: Story = {
  render: args => (
    <Duolist {...args}>
      <DuolistGroup term={'test term'} description={descriptionWithLink} />
      <DuolistGroup term={'test term lang'} description={shortLoremText} />
      <DuolistGroup term={'test term enda lenger'} description={mediumLoremText} />
      <DuolistGroup term={'test term'} description={'test description'} />
    </Duolist>
  ),
};

export const WithLabel: Story = {
  render: args => (
    <Duolist
      {...args}
      label={
        <Title appearance="title3" htmlMarkup="h3">
          {'Test label'}
        </Title>
      }
    >
      <DuolistGroup term={'test term'} description={'kort tekst'} />
      <DuolistGroup term={'test term lang'} description={shortLoremText} />
      <DuolistGroup term={'test term enda lenger'} description={mediumLoremText} />
      <DuolistGroup term={'test term'} description={'test description'} />
    </Duolist>
  ),
};

export const WithLines: Story = {
  args: {
    variant: 'line',
  },
  render: args => (
    <Duolist {...args}>
      <DuolistGroup term={'test term'} description={'kort tekst'} />
      <DuolistGroup term={'test term lang'} description={shortLoremText} />
      <DuolistGroup term={'test term enda lenger'} description={mediumLoremText} />
      <DuolistGroup term={'test term'} description={'test description'} />
    </Duolist>
  ),
};

export const WithBorder: Story = {
  args: {
    border: 'border',
  },
  render: args => (
    <Duolist {...args}>
      <DuolistGroup term={'test term'} description={'kort tekst'} />
      <DuolistGroup term={'test term lang'} description={shortLoremText} />
      <DuolistGroup term={'test term enda lenger'} description={mediumLoremText} />
      <DuolistGroup term={'test term'} description={'test description'} />
    </Duolist>
  ),
};

export const BoldColumn: Story = {
  render: args => (
    <>
      <Duolist {...args}>
        <DuolistGroup term={'test term'} description={'kort tekst'} />
        <DuolistGroup boldColumn={'second'} term={'test term lang'} description={shortLoremText} />
        <DuolistGroup term={'test term enda lenger'} description={mediumLoremText} />
        <DuolistGroup boldColumn={'second'} term={'test term'} description={'test description'} />
      </Duolist>
      <br />
      <br />
      <Duolist boldColumn="second" {...args}>
        <DuolistGroup term={'test term'} description={'kort tekst'} />
        <DuolistGroup term={'test term lang'} description={shortLoremText} />
        <DuolistGroup boldColumn={'first'} term={'test term enda lenger'} description={mediumLoremText} />
        <DuolistGroup term={'test term'} description={'test description'} />
      </Duolist>
    </>
  ),
};

export const DescriptionWidth: Story = {
  render: args => (
    <>
      <Duolist descriptionWidth={40} {...args}>
        <DuolistGroup term={'Her er en test term'} description={mediumLoremText} />
      </Duolist>
      <Duolist descriptionWidth={60} {...args}>
        <DuolistGroup term={'Her er en test term'} description={mediumLoremText} />
      </Duolist>
      <Duolist descriptionWidth={80} {...args}>
        <DuolistGroup term={'Her er en test term'} description={mediumLoremText} />
      </Duolist>
    </>
  ),
};

export const NonFormatted: Story = {
  render: args => (
    <>
      <Duolist format="non-formatted" {...args}>
        <DuolistGroup term={'Her er en test term'} description={mediumLoremText} />
        <DuolistGroup term={'Her er en test term nr 2'} description={mediumLoremText} />
      </Duolist>
      <Duolist format="non-formatted" separator=" - " {...args}>
        <DuolistGroup term={'Custom separator'} description={mediumLoremText} />
      </Duolist>
    </>
  ),
};
