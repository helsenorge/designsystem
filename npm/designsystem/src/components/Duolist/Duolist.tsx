import React from 'react';

import classNames from 'classnames';

import { AnalyticsId } from '../../constants';
import { Breakpoint, useBreakpoint } from '../../hooks/useBreakpoint';
import Spacer from '../Spacer';
import { TitleProps } from '../Title';

import duolistStyles from './styles.module.scss';

export type DuolistVariants = 'normal' | 'line';
export type BoldColumn = 'first' | 'second' | 'none';
export type Border = 'no-border' | 'border';
export type Formats = 'formatted' | 'non-formatted';

export interface DuolistProps {
  /** Determines which column is bold */
  boldColumn?: BoldColumn;
  /** Label of the Duolist */
  border?: Border;
  /** Label of the Duolist */
  label?: React.ReactElement<TitleProps>;
  /** Formatted or non-formatted visual variants */
  format?: Formats;
  /** Character separator for non-formatted format */
  separator?: string;
  /** Sets the visual variant of the Duolist. */
  variant?: DuolistVariants;
  /** Sets the content of the Duolist. */
  children: React.ReactNode;
  /** Adds custom classes to the element. */
  className?: string;
  /** Sets the data-testid attribute. */
  testId?: string;
  /** Width of the description column in percentage */
  descriptionWidth?: number;
  /** Use collapsed mode on columns from chosen breakpoint and below. */
  useCollapsedFromAndBelowBreakpoint?: keyof typeof Breakpoint;
}

export interface DuolistGroupProps {
  /** Determines which column is bold */
  boldColumn?: BoldColumn;
  /** Sets content of the <dd> tag. */
  description: React.ReactNode;
  /** Formatted or non-formatted visual variants */
  format?: Formats;
  /** Character separator for non-formatted format */
  separator?: string;
  /** Sets content of the <dt> tag. */
  term: React.ReactNode;
}

export const DuolistGroup: React.FC<DuolistGroupProps> = props => {
  const { format = 'formatted', boldColumn = format === 'non-formatted' ? 'none' : 'first', description, separator = ': ', term } = props;

  const firstBold = boldColumn === 'first';
  const secondBold = boldColumn === 'second';
  const nonFormatted = format === 'non-formatted';

  const dtClassNames = classNames(duolistStyles['duolist__dt'], {
    [duolistStyles['duolist__dt--bold']]: firstBold,
    [duolistStyles['duolist__dt--non-formatted']]: nonFormatted,
  });
  const ddClassNames = classNames(duolistStyles['duolist__dd'], {
    [duolistStyles['duolist__dd--bold']]: secondBold,
    [duolistStyles['duolist__dd--non-formatted']]: nonFormatted,
  });

  const renderContent = () => {
    return (
      <>
        <dt data-separator={nonFormatted ? separator : undefined} className={dtClassNames}>
          {term}
        </dt>
        <dd className={ddClassNames}>{description}</dd>
      </>
    );
  };

  return nonFormatted ? <div>{renderContent()}</div> : <>{renderContent()}</>;
};

export const Duolist: React.FC<DuolistProps> = props => {
  const {
    boldColumn,
    border = 'no-border',
    descriptionWidth,
    label,
    format = 'formatted',
    separator,
    variant = 'normal',
    children,
    className,
    testId,
    useCollapsedFromAndBelowBreakpoint,
  } = props;

  const hasBorder = border === 'border';
  const hasLines = variant === 'line';
  const extraPaddingTop = hasBorder && (label || hasLines);
  const nonFormatted = format === 'non-formatted';
  const breakpoint = useBreakpoint();
  const useCollapsedMode = useCollapsedFromAndBelowBreakpoint && breakpoint <= Breakpoint[useCollapsedFromAndBelowBreakpoint];

  const duolistWrapperClasses = classNames(
    {
      [duolistStyles['duolist-wrapper--border']]: hasBorder,
      [duolistStyles['duolist-wrapper--extra-padding-top']]: extraPaddingTop,
    },
    className
  );

  const duolistClasses = classNames(duolistStyles.duolist, {
    [duolistStyles['duolist--line']]: hasLines,
    [duolistStyles['duolist--non-formatted']]: nonFormatted,
    [duolistStyles['duolist--collapsed']]: useCollapsedMode,
    [duolistStyles['duolist--not-collapsed']]: !useCollapsedMode,
  });

  const duolistColumnStyle = descriptionWidth ? descriptionWidth + '%' : 'minmax(60%, 1fr)';

  return (
    <div className={duolistWrapperClasses} data-testid={testId} data-analyticsid={AnalyticsId.Duolist}>
      {label && (
        <>
          {label}
          <Spacer />
        </>
      )}
      <dl
        style={!nonFormatted ? { gridTemplateColumns: useCollapsedMode ? `1fr` : `auto ${duolistColumnStyle}` } : undefined}
        className={duolistClasses}
      >
        {React.Children.map(children, (child: React.ReactNode | React.ReactElement<DuolistGroupProps>) => {
          if (child === null || typeof child === 'undefined') return;
          const duolistGroup = child as React.ReactElement<DuolistGroupProps>;
          if (duolistGroup.type === DuolistGroup) {
            return React.cloneElement(child as React.ReactElement<DuolistGroupProps>, {
              boldColumn: duolistGroup.props.boldColumn ?? boldColumn,
              format: duolistGroup.props.format ?? format,
              separator: duolistGroup.props.separator ?? separator,
            });
          }
        })}
      </dl>
    </div>
  );
};

export default Duolist;
