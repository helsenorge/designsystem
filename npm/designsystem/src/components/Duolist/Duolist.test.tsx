import { render, screen } from '@testing-library/react';

import Duolist, { DuolistGroup } from './Duolist';
import AnchorLink from '../AnchorLink';
import Title from '../Title';

describe('Gitt at Duolist skal vises', (): void => {
  describe('Når Duolist skal vises vanlig', (): void => {
    test('Så rendres den som den skal', (): void => {
      const { container } = render(
        <Duolist testId={'test01'}>
          <DuolistGroup term={'test term 1'} description={'test description 1'} />
          <DuolistGroup term={'test term 2'} description={'test description 2'} />
        </Duolist>
      );

      const duoListWrapper = screen.getByTestId('test01');

      const dlGroupTerm1 = screen.getByText('test term 1');
      const dlGroupTerm2 = screen.getByText('test term 2');
      const dlGroupDescription1 = screen.getByText('test description 1');
      const dlGroupDescription2 = screen.getByText('test description 2');

      expect(duoListWrapper).toBeInTheDocument();
      expect(dlGroupTerm1).toBeInTheDocument();
      expect(dlGroupTerm2).toBeInTheDocument();
      expect(dlGroupDescription1).toBeInTheDocument();
      expect(dlGroupDescription2).toBeInTheDocument();

      expect(dlGroupTerm1).toHaveClass('duolist__dt duolist__dt--bold');
      expect(dlGroupDescription1.className).toBe('duolist__dd');

      expect(container).toMatchSnapshot();
    });
  });

  describe('Når Duolist vises med DuolistVariants line', (): void => {
    test('Så rendres den med line class på duoList', (): void => {
      render(
        <Duolist testId={'test01'} variant={'line'}>
          <DuolistGroup term={'test term 1'} description={'test description 1'} />
          <DuolistGroup term={'test term 2'} description={'test description 2'} />
        </Duolist>
      );

      // eslint-disable-next-line testing-library/no-node-access
      const duoList = screen.getByTestId('test01').children[0];

      expect(duoList).toHaveClass('duolist--line');
    });
  });

  describe('Når Duolist vises med border', (): void => {
    test('Så rendres den med border class på duoList', (): void => {
      render(
        <Duolist testId={'test01'} border="border">
          <DuolistGroup term={'test term 1'} description={'test description 1'} />
          <DuolistGroup term={'test term 2'} description={'test description 2'} />
        </Duolist>
      );

      const duoList = screen.getByTestId('test01');

      expect(duoList).toHaveClass('duolist-wrapper--border');
    });
  });

  describe('Når Duolist vises med label', (): void => {
    test('Så rendres den som den skal', (): void => {
      const label = (
        <Title appearance="title3" htmlMarkup="h3">
          {'Test label'}
        </Title>
      );

      render(
        <Duolist label={label}>
          <DuolistGroup term={'test term 1'} description={'test description 1'} />
          <DuolistGroup term={'test term 2'} description={'test description 2'} />
        </Duolist>
      );

      const labelElement = screen.getByText('Test label');

      expect(labelElement).toBeInTheDocument();
    });
  });

  describe('Når Duolist vises med boldColumn satt som second', (): void => {
    test('Så rendres den med bold class på andre column', (): void => {
      render(
        <Duolist boldColumn="second">
          <DuolistGroup term={'test term 1'} description={'test description 1'} />
          <DuolistGroup term={'test term 2'} description={'test description 2'} />
          <DuolistGroup boldColumn={'first'} term={'test term 3'} description={'test description 3'} />
        </Duolist>
      );

      const duoListTerm1 = screen.getByText('test term 1');
      const duoListDescription1 = screen.getByText('test description 1');
      const duoListTerm2 = screen.getByText('test term 2');
      const duoListDescription2 = screen.getByText('test description 2');
      const duoListTerm3 = screen.getByText('test term 3');
      const duoListDescription3 = screen.getByText('test description 3');

      expect(duoListTerm1.className).toBe('duolist__dt');
      expect(duoListDescription1).toHaveClass('duolist__dd duolist__dd--bold');
      expect(duoListTerm2.className).toBe('duolist__dt');
      expect(duoListDescription2).toHaveClass('duolist__dd duolist__dd--bold');
      expect(duoListTerm3.className).toBe('duolist__dt duolist__dt--bold');
      expect(duoListDescription3).toHaveClass('duolist__dd');
    });
  });

  describe('Når Duolist vises med anchorlink', (): void => {
    test('Så rendres det riktig', (): void => {
      render(
        <Duolist testId={'test01'} border="border">
          <DuolistGroup
            term={'test term 1'}
            description={
              <>
                {'test description 1 + '} <AnchorLink>{'test link 1'}</AnchorLink>
              </>
            }
          />
          <DuolistGroup term={'test term 2'} description={'test description 2'} />
        </Duolist>
      );

      const link = screen.getByText('test link 1');

      expect(link).toBeInTheDocument();
      expect(link).toHaveClass('anchorlink');
    });
  });
  describe('Når Duolist child er null', (): void => {
    test('Så kastes det ikke en feilmelding', (): void => {
      render(
        <Duolist testId={'test01'} border="border">
          {false && <DuolistGroup term={'test term 1'} description={'test description 1'} />}
        </Duolist>
      );
    });
  });
  describe('Når Duolist format er satt til non-formatted', (): void => {
    test('Så rendrer den riktig', (): void => {
      render(
        <Duolist testId={'test01'} format={'non-formatted'}>
          <DuolistGroup term={'test term 1'} description={'test description 1'} />
        </Duolist>
      );

      // eslint-disable-next-line testing-library/no-node-access
      const duoList = screen.getByTestId('test01').children[0];
      const dt = screen.getByText('test term 1');
      const dd = screen.getByText('test description 1');
      const separator = dt.getAttribute('data-separator');

      expect(duoList.className).toBe('duolist duolist--non-formatted');
      expect(dt.className).toBe('duolist__dt duolist__dt--non-formatted');
      expect(dd.className).toBe('duolist__dd duolist__dd--non-formatted');
      expect(separator).toBe(': ');
    });
  });

  describe('Når Duolist separator er satt med format non-formatted', (): void => {
    test('Så settes den ønskede separatoren', (): void => {
      render(
        <Duolist testId={'test01'} format={'non-formatted'} separator=" - ">
          <DuolistGroup term={'test term 1'} description={'test description 1'} />
        </Duolist>
      );

      const dt = screen.getByText('test term 1');
      const separator = dt.getAttribute('data-separator');

      expect(separator).toBe(' - ');
    });
  });

  describe('Når Duolist separator er satt med format formatted', (): void => {
    test('Så settes det ikke en separator', (): void => {
      render(
        <Duolist testId={'test01'} separator=" - ">
          <DuolistGroup term={'test term 1'} description={'test description 1'} />
        </Duolist>
      );

      const dt = screen.getByText('test term 1');
      const separator = dt.getAttribute('data-separator');

      expect(separator).toBe(null);
    });
  });
});
