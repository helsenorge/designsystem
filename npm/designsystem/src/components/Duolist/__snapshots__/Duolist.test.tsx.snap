// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Gitt at Duolist skal vises > NÃ¥r Duolist skal vises vanlig > SÃ¥ rendres den som den skal 1`] = `
<div>
  <div
    class=""
    data-analyticsid="duolist"
    data-testid="test01"
  >
    <dl
      class="duolist duolist--not-collapsed"
      style="grid-template-columns: auto minmax(60%, 1fr);"
    >
      <dt
        class="duolist__dt duolist__dt--bold"
      >
        test term 1
      </dt>
      <dd
        class="duolist__dd"
      >
        test description 1
      </dd>
      <dt
        class="duolist__dt duolist__dt--bold"
      >
        test term 2
      </dt>
      <dd
        class="duolist__dd"
      >
        test description 2
      </dd>
    </dl>
  </div>
</div>
`;
